README.md
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/surrosurv)](https://cran.r-project.org/package=surrosurv)
[![Build Status](https://travis-ci.org/Oncostat/surrosurv.svg?branch=master)](https://travis-ci.org/Oncostat/surrosurv)

# surrosurv
Evaluation of Failure Time Surrogate Endpoints in Individual Patient Data Meta-Analyses

## Description
Provides functions for the evaluation of
    surrogate endpoints when both the surrogate and the true endpoint are failure
    time variables. The approaches implemented are: (1) the two-step approach
    (Burzykowski et al, 2001) [<DOI:10.1111/1467-9876.00244>](http://dx.doi.org/10.1111/1467-9876.00244)
    with a copula model (Clayton, Plackett, Hougaard) at
    the first step and either a linear regression of log-hazard ratios at the second
    step (either adjusted or not for measurement error); (2) mixed proportional
    hazard models estimated via mixed Poisson GLM.

NEWS
Changes in version 1.1.23 (2017-07-13)
- new function simData.gh() to generate data from Gumbel-Hougaard copula

Changes in version 1.1.15 (2017-04-25)
- fixed minor issue in prediction function for adjusted copula models

Changes in version 1.1.14 (2017-04-06)
- fixed issue in prediction intervals for mixed Poisson models

Changes in version 1.1.13 (2017-03-28)
- fixed issue in prediction intervals for adjusted copula models

Changes in version 1.1.12 (2017-03-02)
- loocv() now also returns the values of kTau and R2
  estimated in each (N-1) fold

Changes in version 1.1.10 (2017-02-27)
- fixed issues with loocv when few trials (added controls)
- added data 'gastadj'
- added twoStage parameter to surrosurv for Shih and Louis (1995) approach
  to copula estimation

Changes in version 1.1.4 (2016-12-06)
- added paper manuscript as vignette('surrosurv')

Changes in version 1.1 (2016-11-09)
- Poisson models can be fitted each separately

################################################################################

Changes in version 0.1.1 (2016-09-28)
- fixed examples for poissonize()

Changes in version 0.1.0 (First Complete version, 2016-09-27)
- new function loocv() (with print() and plot() functions)
  to compute leave-one-out cross-validation

################################################################################

Changes in version 0.0.15 (2016-09-23)
- new function ste() to compute the surrogate threshold effect
- plot.sussosurv() can now show prediction intervals and the STE

Changes in version 0.0.11 (2016-08-12)
- predict and plot for class sussosurv

Changes in version 0.0.10 (2016-08-01)
- kkt2 convergence criteria corrected from positive determinant
  to positive min eigenvalue

Changes in version 0.0.9 (2016-07-28)
- bugfix in Poisson method, which did not return results
  because of mispelled model name

Changes in version 0.0.7 (2016-07-25)
- the Kendall's tau for copulas is now initialized
  using the SurvCorr package (much faster)

Changes in version 0.0.6 (2016-07-22)
- added the function convals(), giving the values of the max abs gradient
  and the min eigenvalue of the variance-covariance matrix of the random treatment effects
- the function convergence() uses explicit computation provided by covals(),
  instead of using the function optimx in the package optimx

R/predict.R
predict.surrosurv <- function(object,
                              models = names(object),
                              exact.models,
                              ...) {
  if (missing(exact.models))
    exact.models <- all(sapply(tolower(noSpP(models)), function(m)
      any(tolower(noSpP(
        names(object)
      )) %in% m)))
  
  if (exact.models) {
    ind <-
      which(tolower(noSpP(names(object))) %in% tolower(noSpP(models)))
  } else {
    ind <- which(sapply(tolower(noSpP(names(
      object
    ))), function(mod)
      ! all(is.na(
        pmatch(tolower(noSpP(models)), mod)
      ))))
  }
  models <- names(object)[ind]
  
  copulas <- intersect(c('Clayton', 'Plackett', 'Hougaard'), models)
  poissons <- grep('PoissonT', models, value = TRUE)
  
  allRES <- c(# Copula models
    sapply(copulas, function(cop) {
      list(unadj = as.data.frame(object[[cop]]$unadj$step1[c('alpha', 'beta')]),
           adj = as.data.frame(object[[cop]]$adj$ranef))
    }),
    # Poisson models
    lapply(poissons, function(poi) {
      # object[[poi]]$ranef$trialref[, c('trtS', 'trtT')]
      as.data.frame(t(t(object[[poi]]$ranef$trialref[, c('trtS', 'trtT')]) +
                        unlist(object[[poi]][c('alpha', 'beta')])))
    }))
  
  allRES <- lapply(allRES, function(object) {
    colnames(object) <- c('trtS', 'trtT')
    return(object)
  })
  names(allRES) <- c(paste(rep(copulas, each = 2), rep(c('unadj', 'adj'), length(copulas)), sep =
                             '.'),
                     poissons)
  
  class(allRES)  <- c('predictSurrosurv', class(allRES))
  attr(allRES, 'trialSizes') <- attr(object, 'trialSizes')
  
  attr(allRES, 'predf') <- c(# Copula models
    sapply(copulas, function(cop) {
      list(unadj = Vectorize(function(x) {
        # est <- coef(object[[cop]]$unadj$step2) %*% c(1, x)
        # n <- 2 + object[[cop]]$unadj$step2$df.residual
        # t <- qt(c(.025, .975), n - 2)
        # S <- summary(object[[cop]]$unadj$step2)$sigma
        # sqrt(
        #     1 + 1 / n + n * (
        #       x - mean(object[[cop]]$unadj$step1$alpha)
        #     )^2 / (
        #       n * sum(object[[cop]]$unadj$step1$alpha^2) -
        #         sum(object[[cop]]$unadj$step1$alpha)^2
        #     )
        # ) * S * t + est
        res <- as.vector(predict(object[[cop]]$unadj$step2,
                                 data.frame(alpha = x),
                                 interval = 'prediction'))
        names(res) <- c('fit', 'lwr', 'upr')
        return(res)
      }),
      adj = Vectorize(function(x) {
        cMEAN <- object[[cop]]$adj$step2$gamma %*% c(1, x)
        # The following variance is given by equation (18.27), page 332
        # Burzykowski and Buyse, An alternative measure for surrogate
        # endpoint validation, in Burzykowski, Molenberghs, Buyse, 2005
        # The evaluation of surrogate endpoints, New York Springer
        # library(Matrix)
        pdVCOV <- nearPD(object[[cop]]$adj$step2$gamma.vcov)$mat
        pVAR <-
          pdVCOV[1, 1] + 2 * x * pdVCOV[1, 2] + x ^ 2 * pdVCOV[2, 2] +
          object[[cop]]$adj$step2$sigma
        res <- rep(cMEAN, 3) +
          qnorm(c(
            fit = .5,
            lwr = .025,
            upr = .975
          )) * sqrt(pVAR)
        names(res) <- c('fit', 'lwr', 'upr')
        return(res)
      }))
    }),
    # Poisson models
    lapply(poissons, function(poi) {
      Vectorize(function(x) {
        # VCOV <- as.matrix(
        #   object[[poi]]$VarCor$trialref[paste0('trt', c('S', 'T')),
        #                                 paste0('trt', c('S', 'T'))] +
        #     object[[poi]]$fixVarCor[paste0('trt', c('S', 'T')),
        #                             paste0('trt', c('S', 'T'))])
        # CORR <- VCOV[1, 2] / sqrt(prod(diag(VCOV)))
        # AVG <- unlist(object[[poi]][c('alpha', 'beta')])
        # cMEAN <- AVG['beta.trtT'] + (x - AVG['alpha.trtS']) * CORR *
        #   # sqrt(VCOV['trtT', 'trtT'] / VCOV['trtS', 'trtS'])
        #   sqrt(exp(diff(log(diag(VCOV)))))
        # res <- as.vector(
        #   rep(cMEAN, 3) + qnorm(c(fit=.5, lwr=.025, upr=.975)) *
        #     sqrt(VCOV['trtS', 'trtS'] * (1 - VCOV[1, 2] /
        #                                    prod(sqrt(diag(VCOV))))))
        cMEAN <- object[[poi]]$regr$gamma %*% c(1, x)
        # The following variance is given by equation (18.27), page 332
        # Burzykowski and Buyse, An alternative measure for surrogate
        # endpoint validation, in Burzykowski, Molenberghs, Buyse, 2005
        # The evaluation of surrogate endpoints, New York Springer
        # library(Matrix)
        pdVCOV <- nearPD(object[[poi]]$regr$gamma.vcov)$mat
        pVAR <-
          pdVCOV[1, 1] + 2 * x *  pdVCOV[1, 2] + x ^ 2 * pdVCOV[2, 2] +
          object[[poi]]$regr$sigma
        res <- rep(cMEAN, 3) +
          qnorm(c(
            fit = .5,
            lwr = .025,
            upr = .975
          )) * sqrt(pVAR)
        names(res) <- c('fit', 'lwr', 'upr')
        return(res)
      })
    }))
  attr(allRES, 'surro.stats') <- print(object, silent = TRUE)
  attr(allRES, 'surro.stats') <- attr(allRES, 'surro.stats')[noSpP(rownames(attr(allRES, 'surro.stats'))) %in%
                                                               noSpP(names(allRES)), , drop = FALSE]
  noSpP(names(allRES))
  names(attr(allRES, 'predf')) <- names(allRES)
  return(allRES)
}

format.methodNames <- function(x) {
  sub(
    '.unadj',
    ' copula (Unadjusted)',
    sub(
      '.adj',
      ' copula (Adjusted)',
      sub('Poisson', 'Poisson ', names(x)),
      fixed = TRUE
    ),
    fixed = TRUE
  )
}

noSpP <- function(x)
  gsub('[\\. ]', '', x)

print.predictSurrosurv <- function(x, n = 6, ...) {
  cat('Treatment effect prediction for surrosurv object\n')
  for (i in 1:length(x)) {
    method <- format.methodNames(x)[i]
    cat('\n  ', method, '\n')
    res2print <- format(t(head(x[[i]], n)), digits = 1)
    if (nrow(x[[i]] > n))
      res2print <- cbind(res2print, '  ' = c('...', '...'))
    rownames(res2print) <- paste0(sub('trt', '    Treatment effects on ', rownames(res2print)), ':')
    print(res2print, quote = FALSE, ...)
  }
}

##############################################################################

ste <- function(x, models = names(x), exact.models) {
  if (class(x)[1] == 'surrosurv')
    x <- predict(x)
  if (missing(exact.models))
    exact.models <- all(sapply(tolower(noSpP(models)), function(m)
      any(tolower(noSpP(
        names(x)
      )) %in% m)))
  
  if (exact.models) {
    ind <- which(tolower(noSpP(names(x))) %in% tolower(noSpP(models)))
  } else {
    ind <- which(sapply(tolower(noSpP(names(
      x
    ))), function(mod)
      ! all(is.na(
        pmatch(tolower(noSpP(models)), mod)
      ))))
  }
  
  res <- sapply(ind, function(i) {
    f <- function(y)
      attr(x, 'predf')[[i]](y)['upr', ] ^ 2
    ste <- optimize(f, c(-1e8, 1e8))$minimum
    ste <- ifelse(abs(f(ste)) > 1e-4, NA, ste)
    return(ste)
  })
  names(res) <- names(x)[ind]
  class(res) <- 'steSurrosurv'
  return(res)
}

print.steSurrosurv <- function(x, digits = 2, ...) {
  res2print <- cbind(beta = round(x, digits),
                     HR = round(exp(x), digits))
  print(res2print, quote = FALSE, ...)
}


##############################################################################

plot.surrosurv <- function(x, ...) {
  plot(predict(x), ...)
}


plot.predictSurrosurv <- function(x,
                                  models = names(x),
                                  exact.models,
                                  pred.ints = TRUE,
                                  show.ste = TRUE,
                                  surro.stats = TRUE,
                                  xlab,
                                  ylab,
                                  xlim,
                                  ylim,
                                  mfrow,
                                  main,
                                  ...) {
  # ************************************************************************** #
  if (missing(xlab))
    xlab <- 'Treatment effect (HR) on S'
  if (missing(ylab))
    ylab <- 'Treatment effect (HR) on T'
  if (missing(exact.models))
    exact.models <- all(sapply(tolower(noSpP(models)), function(m)
      any(tolower(noSpP(
        names(x)
      )) %in% m)))
  w <- attr(x, 'trialSizes')
  if (var(w)) {
    w <- .8 + 3 * (w - min(w)) / (max(w) - min(w))
  } else
    w <- 1
  
  if (exact.models) {
    ind <- which(tolower(noSpP(names(x))) %in% tolower(noSpP(models)))
  } else {
    ind <- which(sapply(tolower(noSpP(names(
      x
    ))), function(mod)
      ! all(is.na(
        pmatch(tolower(noSpP(models)), mod)
      ))))
  }
  PREDF <- attr(x, 'predf')[ind]
  if (show.ste)
    STE <- ste(x, names(x)[ind], exact.models = TRUE)
  if (surro.stats) {
    SURRO.STATS <-
      round(matrix(as.numeric(attr(
        x, 'surro.stats'
      )), ncol = 2),
      2)[ind, , drop = FALSE]
    SURRO.STATS[is.na(SURRO.STATS)] <- '\u2014'
  }
  x <- x[ind]
  
  if (!missing('xlim'))
    xlims <- log(xlim)
  if (!missing('ylim'))
    ylims <- log(ylim)
  
  if (length(x)) {
    if (missing(mfrow)) {
      par(mfrow = n2mfrow(length(x)))
    } else {
      par(mfrow = mfrow)
    }
    for (i in 1:length(x)) {
      # abcoeff <- try(coef(lm(x[[i]][, 2:1])), silent = TRUE)
      predf.fit <- Vectorize(function(y) {
        PREDF[[i]](y)[1]
      })
      set0in <- function(xint) {
        if (xint[1] > 0)
          xint[1] <- 0
        if (xint[2] < 0)
          xint[2] <- 0
        five <- diff(range(xint)) / 20
        return(xint + c(-1, 1) * five)
      }
      if (missing('xlim'))
        xlims <- set0in(range(x[[i]][, 1]))
      if (missing('ylim'))
        ylims <- set0in(range(x[[i]][, 2]))
      plot(
        x[[i]],
        asp = 1,
        xlim = xlims,
        ylim = ylims,
        cex = w,
        pch = 21,
        bg = rgb(.5, .5, .5, .5),
        panel.first = {
          if (pred.ints) {
            Xs <- seq(axTicks(1)[1] - 1, rev(axTicks(1))[1] + 1,
                      length.out = 1e2)
            polygon(
              c(Xs, rev(Xs)),
              c(PREDF[[i]](Xs)[2,], rev(PREDF[[i]](Xs)[3,])),
              col = rgb(.8, .8, .8, .75),
              border = NA
            )
          }
          abline(h = 0, v = 0, col = 'grey')
          # if (all(is.finite(abcoeff)))
          # abline(lm(x[[i]][, 2:1]), col=rgb(.2, .2, .2, .8), lwd = 2)
          curve(
            predf.fit,
            from = axTicks(1) - 1,
            to = rev(axTicks(1))[1] + 1,
            col = rgb(.2, .2, .2, .8),
            lwd = 2,
            add = TRUE
          )
        },
        xaxt = 'n',
        yaxt = 'n',
        main =  ifelse(missing(main), format.methodNames(x)[i], main),
        xlab = xlab,
        ylab = ylab,
        ...
      )
      if (show.ste) {
        if (is.na(STE[i])) {
          message(paste0(
            '[', format.methodNames(x)[i], 
            '] The STE does not exist: the upper limits ',
            'of the prediction intervals never cross the line HR=1'))
        } else {
          points(STE[i],
                 0,
                 col = 2,
                 pch = '|',
                 font = 2)
          text(
            STE[i],
            0,
            col = 2,
            font = 2,
            adj = c(1.1, -.5),
            labels = paste('STE =', round(exp(STE[i]), 2))
          )
        }
      }
      if (surro.stats) {
        mtext(
          bquote(paste(R ^ 2, ' = ', .(
            SURRO.STATS[i, 2]
          ))),
          side = 1,
          line = -1.2,
          adj = .9,
          font = 2
        )
        mtext(
          bquote(paste(tau, ' = ', .(
            SURRO.STATS[i, 1]
          ))),
          side = 1,
          line = -2.8,
          adj = .9,
          font = 2
        )
      }
      axis(1,
           at = axTicks(1),
           labels = sapply(exp(axTicks(1)), function(x)
             ifelse(
               x < .1,
               format(x, nsmall = 3, digits = 2),
               ifelse(
                 x < 1,
                 format(x, nsmall = 2, digits = 2),
                 ifelse(
                   x < 10,
                   format(x, nsmall = 1, digits = 2),
                   format(x, digits = 2, drop0trailing =
                            FALSE)
                 )
               )
             )))
      axis(2,
           at = axTicks(2),
           labels = sapply(exp(axTicks(2)), function(x)
             ifelse(
               x < .1,
               format(x, nsmall = 3, digits = 2),
               ifelse(
                 x < 1,
                 format(x, nsmall = 2, digits = 2),
                 ifelse(
                   x < 10,
                   format(x, nsmall = 1, digits = 2),
                   format(x, digits = 2, drop0trailing =
                            FALSE)
                 )
               )
             )))
    }
  } else
    warning('No model selected', immediate. = TRUE)
}

R/convergence.R
# ---------------------------------------------------------------------------- #
convals <- function(x, ...) UseMethod('convals')
# ---------------------------------------------------------------------------- #

# ---------------------------------------------------------------------------- #
convals.surrosurv <- function(x, ...) {
  models <- c('Clayton', 'Plackett', 'Hougaard', 'Poisson')
  models <-
    models[sapply(models, function(y)
      any(grepl(y, names(x))))]
  
  copulas <- function(cop) {
    f <- function() {
      res <- t(sapply(x[[cop]], function(y) {
        vals <- c(maxSgrad = tryCatch(
          max(abs(with(
            attributes(y$optimxRES), solve(hessian, grad)
          ))),
          error = function(x)
            return(Inf)
        ),
        minHev = tryCatch(
          min(eigen(attr(
            y$optimxRES, 'hessian'
          ))$values),
          error = function(x)
            return(-1)
        ),
        minREev = tryCatch(
          min(eigen(y$VarCor2)$values),
          error = function(x)
            return(-1)
        ))
        return(vals)
      }))
      rownames(res) <- paste(cop, rownames(res))
      return(res)
    }
    return(f)
  }
  Clayton = copulas('Clayton')
  Plackett = copulas('Plackett')
  Hougaard = copulas('Hougaard')
  Poisson = function() {
    which_models <- x[grepl('Poisson', names(x))]
    convres <- t(sapply(which_models,
                        function(y)
                          tryCatch({
                            vals <- c(maxSgrad = tryCatch(
                              max(abs(with(
                                y$optinfo$derivs, solve(Hessian, gradient)
                              ))),
                              error = function(x)
                                return(Inf)
                            ),
                            minHev = tryCatch(
                              min(eigen(y$optinfo$derivs$Hessian)$values),
                              error = function(x)
                                return(-1)
                            ),
                            minREev = tryCatch(
                              min(eigen(y$VarCor$trialref)$values),
                              error = function(x)
                                return(-1)
                            ))
                            return(vals)
                          }, error = function(x)
                            return(c(
                              Inf, -1, -1
                            )))))
    names(convres) <- names(which_models)
    return(convres)
  }
  
  fitRES <-
    do.call(rbind, lapply(models, function(x)
      eval(call(paste(
        x
      )))))
  fitRES[grepl('unadj|PoissonI', rownames(fitRES)), 'minREev'] <- NA
  class(fitRES) <- 'conv'
  return(fitRES)
}
# ---------------------------------------------------------------------------- #


# ---------------------------------------------------------------------------- #
convergence <- function(x, kkttol = 1e-2, kkt2tol = 1e-8, ...)
  UseMethod('convergence')
# ---------------------------------------------------------------------------- #


# ---------------------------------------------------------------------------- #
convergence.surrosurv <- function(x, kkttol = 1e-2, kkt2tol = 1e-8, ...) {
  ConvRES <- convals(x)
  checkConvRES <- cbind(ConvRES[, 1, drop = FALSE] <= kkttol,
                        ConvRES[, 2:3] > kkt2tol)
  
  class(checkConvRES) <- 'conv'
  return(checkConvRES)
}
# ---------------------------------------------------------------------------- #


# ---------------------------------------------------------------------------- #
print.conv <- function(x, na.print = '---', ...) {
  class(x) <- 'matrix'
  print(x, na.print = na.print, ...)
}
# ---------------------------------------------------------------------------- #

R/loocv.R
################################################################################
loocv <- function(object, ...)
  UseMethod('loocv')
################################################################################
loocv.surrosurv <- function(object,
                            models,
                            nCores,
                            parallel = TRUE,
                            ...) {
  if (missing(models)) {
    allmodels <- c(
      'Clayton',
      'Plackett',
      'Hougaard',
      'Poisson I',
      'Poisson T',
      'Poisson TI',
      'Poisson TIa'
    )
    models <- allmodels[allmodels %in% names(object)]
  }
  loocv.data.frame(attr(object, 'data'), models = models, ...)
}
################################################################################
loocv.data.frame <- function(object,
                             models = c('Clayton', 'Poisson TI'),
                             nCores,
                             parallel = TRUE,
                             ...) {
  # ************************************************************************** #
  models <- tolower(noSpP(models))
  if ('poisson' %in% models) {
    models <- setdiff(models, 'poisson')
    models <-
      unique(c(models, paste0('poisson', c(
        'i', 't', 'ti', 'tia'
      ))))
  }
  models <- match.arg(
    models,
    several.ok = TRUE,
    choices = c('clayton', 'plackett', 'hougaard', paste0('poisson', c(
      'i', 't', 'ti', 'tia'
    )))
  )
  object$trialref <- factor(object$trialref)
  
  nTrials <- nlevels(object$trialref)
  if (nTrials < 3)
    stop('At least three trials are needed for cross-validation.')
  # library('parallel')
  
  if (parallel) {
    totCores <- detectCores()
    
    if (missing(nCores))
      nCores <- NA
    if (is.na(as.integer(nCores))) {
      nCores <- min(nTrials, totCores)
      message(
        paste0(
          'Parallel computing on ',
          nCores,
          ' cores (the total number of ',
          ifelse(nTrials < totCores, 'trials', 'cores detected'),
          ')'
        )
      )
    } else {
      if (nCores > min(nTrials, totCores))
        message(
          paste0(
            'The number of cores (nCores=',
            nCores,
            ') is greater than',
            'the number of ',
            ifelse(nTrials < totCores, 'trials', 'cores detected'),
            ' (',
            min(nTrials, totCores),
            ')'
          )
        )
      
      nCores <- min(nCores, nTrials, totCores)
      message(paste('Parallel computing on', nCores, 'cores'))
    }
  } else
    nCores <- 1
  
  loof <- function(TRIAL, models2predict = models, ...) {
    alpha <- coef(coxph(Surv(timeS, statusS) ~ trt,
                        data = object[object$trialref == TRIAL,]))[['trt']]
    beta <- coef(coxph(Surv(timeT, statusT) ~ trt,
                       data = object[object$trialref == TRIAL,]))[['trt']]
    redobject <- object[object$trialref != TRIAL,]
    redobject$trialref <- factor(redobject$trialref)
    predint <- tryCatch(
      expr = {
        surrofit <- surrosurv(data = redobject, models = models2predict, ...)
        res <- lapply(attr(predict(surrofit), 'predf'), function(x) x(alpha))
        names <- names(res)
        res <- lapply(1:length(res), function(i)
          rbind(res[[i]], t(print(
            surrofit, silent = TRUE
          ))[, i, drop = FALSE]))
        names(res) <- names
        attr(res, 'convergence') <- convergence(surrofit)
        attr(res, 'convals') <- convals(surrofit)
        res
      },
      error = function(e) {
        for (cop in c('clayton', 'plackett', 'hougaard')) {
          coppos <- which(models2predict == cop)
          if (length(coppos) == 1) {
            models2predict <- c(models2predict[1:coppos - 1],
                                paste(cop, c('unadj', 'adj'), sep = '.'),
                                as.character(na.omit(models2predict[
                                  (coppos + 1):(length(models2predict) + 1)])))
          }
        }
        poipos <- which(models2predict == 'Poisson')
        if (length(poipos) == 1) {
          models2predict <- c(models2predict[0:(poipos - 1)],
                              paste0('Poisson', c('T', 'TI', 'TIa')))
        }
        res <- mapply(function(i)
          return(t(t(c(fit = NA,
                       lwr = NA,
                       upr = NA,
                       kTau = NA,
                       R2 = NA)
          ))),
          models2predict, SIMPLIFY = FALSE)
        attr(res, 'convergence') <- attr(res, 'convals') <- 
          matrix(NA, length(models2predict), 3)
        return(res)
      }
    )
    RES <- c(list(margPars = c(alpha = alpha, beta = beta)),
             predint)
    attr(RES, 'convergence') <- attr(predint, 'convergence')
    attr(RES, 'convals') <- attr(predint, 'convals')
    return(RES)
  }
  
  if (Sys.info()[1] == "Windows") {
    cl <- makeCluster(nCores, type = 'PSOCK')
    clusterExport(cl, 'object', environment())
  } else {
    cl <- makeCluster(nCores, type = 'FORK')
  }
  clusterEvalQ(cl, library('survival'))
  clusterEvalQ(cl, library('surrosurv'))
  loocvRES <- clusterApplyLB(cl, levels(object$trialref), loof, ...)
  stopCluster(cl)
  rm(cl)
  
  names(loocvRES) <- levels(object$trialref)
  class(loocvRES) <- c('loocvSurrosurv', class(loocvRES))
  return(loocvRES)
}
################################################################################


################################################################################
print.loocvSurrosurv <- function(x,
                                 n = min(length(x), 6),
                                 silent = FALSE,
                                 ...) {
  # ************************************************************************** #
  models <- setdiff(names(x[[1]]), 'margPars')
  RES <- lapply(models, function(y) {
    preds <- sapply(x, function(trial) {
      trialRes <- if (is.null(trial[[y]])) {
        rep(NA, 5)
      } else {
        trial[[y]]
      }
      c(obsAlpha = trial$margPars['alpha'], 
        obsBeta = trial$margPars['beta'], 
        trialRes)
    })
    preds <- matrix(unlist(preds),
                    nrow = 7,
                    dimnames = list(
                      c('obsAlpha', 'obsBeta', 'predict', 'lwr', 'upr', 'kTau', 'R2'),
                      names(x)
                    ))
    return(preds)
  })
  names(RES) <- models
  
  if (silent)
    return(RES)
  
  n <- min(n, ncol(RES[[1]]))
  for (i in 1:length(RES)) {
    method <- format.methodNames(RES)[i]
    cat('\n  ', method, '\n')
    res2print <-
      format(
        as.data.frame(RES[[i]][, 1:n, drop = FALSE]),
        digits = 1,
        nsmall = 2,
        na.encode = FALSE
      )
    if (ncol(RES[[i]]) > n)
      res2print <-
      cbind(res2print, '  ' = rep('...', nrow(res2print)))
    # rownames(res2print) <- paste0(
    #   sub('trt', '    Treatment effects on ', rownames(res2print)), ':')
    print(res2print, quote = FALSE, ...)
  }
}
################################################################################


################################################################################
plot.loocvSurrosurv <- function(x,
                                models,
                                exact.models, 
                                plot.type = c('classic', 'regression'),
                                main, ylab, xlab, ...) {
  # ************************************************************************** #
  x <- print(x, silent = TRUE)
  
  if (missing(models))
    models <- setdiff(names(x), 'margPars')
  
  if (missing(exact.models))
    exact.models <-
      any(tolower(noSpP(names(x))) %in% tolower(noSpP(models)))
  
  if (exact.models) {
    ind <- which(tolower(noSpP(names(x))) %in% tolower(noSpP(models)))
  } else {
    ind <- which(sapply(tolower(noSpP(names(
      x
    ))), function(mod)
      !all(is.na(
        pmatch(tolower(noSpP(models)), mod)
      ))))
  }
  
  plot.type <- match.arg(plot.type)
  
  mypalette <-
    c(
      orange = rgb(234, 104, 14, maxColorValue = 255, alpha = 180),
      blue = rgb(28, 170, 155, maxColorValue = 255, alpha = 180),
      magenta = rgb(191,  22, 120, maxColorValue = 255, alpha = 180),
      green = rgb(
        184 * .7,
        201,
        27 * .5,
        maxColorValue = 255,
        alpha = 180
      ),
      grey = rgb(79, 71, 68, maxColorValue = 255, alpha = 180)
    )
  palette(mypalette[c(2:4, 1, 5)])
  
  if (length(ind)) {
    par(mfrow = n2mfrow(length(ind)))
    if (!missing(main)) {
      if (length(main) == 1) mains <- rep(main, length(ind))
      else mains <- main
    } else mains <- format.methodNames(x)
    
    if (plot.type == 'classic') {
      XLAB <- ifelse(missing(xlab), 'Trials', xlab)
    } else {
      XLAB <- ifelse(missing(xlab), 'Hazard ratio on the surrogate endpoint', xlab)
    }
    
    for (i in ind) {
      if (plot.type == 'classic') {
        xlims <- c(0, ncol(x[[i]])) + .5
        Xs <- 1:ncol(x[[i]])
      } else {
        xlims <- range(x[[i]]['obsAlpha', ])
        Xs <- x[[i]]['obsAlpha', ]
      }
      plot(
        NA,
        main = mains[i],
        xlim = xlims,
        xlab = XLAB,
        xaxt = 'n',
        ylim = range(x[[i]], na.rm = TRUE) +
          c(-1, 1) * diff(range(x[[i]], na.rm = TRUE)) / 20,
        ylab = ifelse(missing(ylab), 'Hazard ratio on the true endpoint', ylab),
        yaxt = 'n',
        panel.first = abline(h = 0, col = 'grey')
      )
      if (plot.type == 'classic') {
        axis(1, 1:ncol(x[[i]]), labels = colnames(x[[i]]))
      } else axis(1, axTicks(1), format(round(exp(axTicks(1)), 2)), las = 1)
      axis(2, axTicks(2), format(round(exp(axTicks(2)), 2)), las = 1)
      segments(
        Xs,
        x[[i]]['lwr',],
        y1 = x[[i]]['upr',],
        col = rgb(.6, .6, .6, .8),
        lwd = 5
      )
      # segments(1:ncol(x[[i]]) - .01, x[[i]]['predict', ], 1:ncol(x[[i]]) + .01,
      #          col = rgb(.2, .2, .2), lwd = 3, lend = 2)
      points(
        Xs,
        x[[i]]['predict',],
        pch = 15,
        col = rgb(.2, .2, .2),
        lwd = 3,
        lend = 2
      )
      COLs <-
        2 - (colSums(apply(x[[i]][c('obsBeta', 'lwr', 'upr'),], 2,
                           function(x)
                             order(x) == c(2, 1, 3))) == 3)
      NC <- is.na(x[[i]]['lwr',]) | is.na(x[[i]]['upr',])
      COLs[NC] <- 0
      points(Xs,
             x[[i]]['obsBeta',],
             pch = 16,
             cex = 1.4,
             col = COLs)
      if (any(NC))
        mtext('x',
              1,
              -1,
              at = which(NC),
              col = 2,
              font = 2)
    }
  }
}
################################################################################